src/langgraph_google_alloydb_pg/engine.py

```diff
@@ -0,0 +1,521 @@
+            region,
+            cluster,
+            instance,
+            database,
+            user,
+            password,
+            ip_type,
+            iam_account_email=iam_account_email,
+        )
+        return await asyncio.wrap_future(future)
+
+    @classmethod
+    def from_engine(
+        cls: Type[AlloyDBEngine],
+        engine: AsyncEngine,
+        loop: Optional[asyncio.AbstractEventLoop] = None,
+    ) -> AlloyDBEngine:
+        """Create an AlloyDBEngine instance from an AsyncEngine."""
+        return cls(cls.__create_key, engine, loop, None)
+
+    @classmethod
+    def from_engine_args(
+        cls,
+        url: Union[str | URL],
+        **kwargs: Any,
+    ) -> AlloyDBEngine:
+        """Create an AlloyDBEngine instance from arguments
+
+        Args:
+            url (Optional[str]): the URL used to connect to a database. Use url or set other arguments.
+
+        Raises:
+            ValueError: If not all database url arguments are specified
+
+        Returns:
+            AlloyDBEngine
+        """
+        # Running a loop in a background thread allows us to support
+        # async methods from non-async environments
+        if cls._default_loop is None:
+            cls._default_loop = asyncio.new_event_loop()
+            cls._default_thread = Thread(
+                target=cls._default_loop.run_forever, daemon=True
+            )
+            cls._default_thread.start()
+
+        driver = "postgresql+asyncpg"
+        if (isinstance(url, str) and not url.startswith(driver)) or (
+            isinstance(url, URL) and url.drivername != driver
+        ):
+            raise ValueError("Driver must be type 'postgresql+asyncpg'")
+
+        engine = create_async_engine(url, **kwargs)
+        return cls(cls.__create_key, engine, cls._default_loop, cls._default_thread)
+
+    async def _run_as_async(self, coro: Awaitable[T]) -> T:
+        """Run an async coroutine asynchronously"""
+        # If a loop has not been provided, attempt to run in current thread
+        if not self._loop:
+            return await coro
+        # Otherwise, run in the background thread
+        return await asyncio.wrap_future(
+            asyncio.run_coroutine_threadsafe(coro, self._loop)
+        )
+
+    def _run_as_sync(self, coro: Awaitable[T]) -> T:
+        """Run an async coroutine synchronously"""
+        if not self._loop:
+            raise Exception(
+                "Engine was initialized without a background loop and cannot call sync methods."
+            )
+        return asyncio.run_coroutine_threadsafe(coro, self._loop).result()
+
+    async def _ainit_checkpoint_table(
+        self, schema_name: str = "public"
+    ) -> None:
+        """
+        Create AlloyDB tables to save checkpoints.
+
+        Args:
+            schema_name (str): The schema name to store the checkpoint tables.
+                Default: "public".
+
+        Returns:
+            None
+        """
+        MIGRATIONS = [
```

Comment:
I am investigating if we can do better in terms of limiting the number of tables we have and using SQL commands to pull the right data.

Reply:
Of course, I will also look for options to limit the number of tables. If you come to a conclusion, please let me know.